python/taichi/lang/struct.py

```diff
@@ -222,10 +222,20 @@ def fill(self, val):
         Args:
             val (Union[int, float]): Value to fill.
         """
-        def assign_renamed(x, y):
-            return ops.assign(x, y)
 
-        return self._element_wise_writeback_binary(assign_renamed, val)
+        bcast_val = self._broadcast_copy(val)
+        entries = {}
+        for k, v in self.items:
+            if isinstance(v, impl.Expr) and v.ptr.is_tensor():
+                from taichi.lang import matrix_ops  # pylint: disable=C0415
+                matrix_ops.fill(v, val)
+                entries[k] = v
+            elif isinstance(v, (Struct, Matrix)):
+                entries[k] = v._element_wise_binary(ops.assign,
+                                                    bcast_val.entries[k])
+            else:
+                entries[k] = ops.assign(v, bcast_val.entries[k])
+        return Struct(entries)
```

Comment:
The original `fill()` seems a bit problematic. `fill()` should happen in-place and no return value is needed.